### ISSUE TITLE
"AttributeError: 'str' object has no attribute '__dict__'" when Python 3 unpickles Python 2 doctrees

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1070]
When there are cached doctrees that were generated by Sphinx+Python 2.X, and then Sphinx+Python 3.X tries to unpickle them, sometimes it fails with:

```
AttributeError: 'str' object has no attribute '__dict__'
```

While this is a user error, it would be nice if Sphinx handled it more gracefully.

How to reproduce:

```
$ wget -nv http://pypi.python.org/packages/source/d/django-registration/django-registration-0.8.tar.gz
2013-01-08 21:28:29 URL:http://pypi.python.org/packages/source/d/django-registration/django-registration-0.8.tar.gz [284427/284427] -> "django-registration-0.8.tar.gz" [1]

$ tar -xzf django-registration-0.8.tar.gz

$ cd django-registration-0.8/docs/

$ python /usr/local/bin/sphinx-build -q . tmp/
Making output directory...
/home/jwilk/django-registration-0.8/docs/views.rst:52: WARNING: Inline strong start-string without end-string.
WARNING: html_static_path entry '/home/jwilk/django-registration-0.8/docs/_static' does not exist

$ python3 /usr/local/bin/sphinx-build -q . tmp/

Exception occurred:
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/environment.py", line 1215, in get_doctree
    doctree = pickle.load(f)
AttributeError: 'str' object has no attribute '__dict__'
The full traceback has been saved in /tmp/sphinx-err-5j9dc9.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
Either send bugs to the mailing list at <http://groups.google.com/group/sphinx-dev/>,
or report them in the tracker at <https://bitbucket.org/birkenfeld/sphinx/issues/>. Thanks!

$ cat /tmp/sphinx-err-5j9dc9.log
# Sphinx version: 1.1.3
# Python version: 3.2.3
# Docutils version: 0.10 release
# Jinja2 version: 2.6
Traceback (most recent call last):
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/cmdline.py", line 189, in main
    app.build(force_all, filenames)
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/application.py", line 204, in build
    self.builder.build_update()
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/builders/__init__.py", line 196, in build_update
    'out of date' &#37; len(to_build))
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/builders/__init__.py", line 252, in build
    self.write(docnames, list(updated_docnames), method)
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/builders/__init__.py", line 291, in write
    doctree = self.env.get_and_resolve_doctree(docname, self)
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/environment.py", line 1230, in get_and_resolve_doctree
    doctree = self.get_doctree(docname)
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.1.3-py3.2.egg/sphinx/environment.py", line 1215, in get_doctree
    doctree = pickle.load(f)
AttributeError: 'str' object has no attribute '__dict__'
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1070
- Originally reported by: Jakub Wilk
- Originally created at: 2013-01-08T21:48:52.625

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 10:12:19+00:00_

Thanks for the report.

> While this is a user error, it would be nice if Sphinx handled it more gracefully.

I think it's more of an error with Sphinx since it's not invalidating the cached environment if the unpickling fails.

In this case, the `environment.pickle` generated under python2 is being successfully unpickled under python3 but the unpickling of individual doctrees is failing.  

I propose we add additional version information to the saved environment to help with the cache invalidation.  Something like this (needs refactoring):

```
diff --git a/sphinx/environment.py b/sphinx/environment.py
--- a/sphinx/environment.py
+++ b/sphinx/environment.py
@@ -71,8 +71,16 @@ default_settings = {

 # This is increased every time an environment attribute is added
 # or changed to properly invalidate pickle files.
-ENV_VERSION = 42

+import sys
+import docutils
+import sphinx
+
+ENV_VERSION = '''\
+Sphinx: %s
+Docutils: %s
+Python: %s
+''' % (sphinx.__version__, docutils.__version__, sys.version)

 default_substitutions = set([
     'version',
```

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-01-09 10:20:11+00:00_

I'd rather keep compatibility between Sphinx versions where the pickle format hasn't changed, and Python minor versions.

But it is of course possible to do something like

```
#!python
ENV_VERSION = 42 + (sys.version_info[0] - 2)
```

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 14:27:48+00:00_

I suspect that there would be unexpected issues when using saved
doctrees created under a different version of Docutils.

The Python minor version is less of an issue than the Sphinx or
Docutils version except with regards to autodoc.  A different
Python version may indicate the virtualenv or the load path was
specifically changed in order to make the documented objects
resolve differently.

How often are people building the same set of docs using multiple
versions of Sphinx/Docutils/Python anyway?  Is forcing a fresh
build in these cases that big a deal?

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-01-09 14:39:47+00:00_

I don't know; probably not, but there are some pretty big projects out there, and I don't want to make their life harder than it has to be.

Regarding sphinx, every time something is changed so that doctrees or env unpickle would fail, the ENV_VERSION should be increased.

Regarding docutils, although they are known for API breakage between versions, they have (so far) refrained from removing any node classes, which is about the only thing I can see would break unpickling.

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 14:42:34+00:00_

Okay.  

I'll go with the approach you mentioned, ie:

```
ENV_VERSION = 42 + (sys.version_info[0] - 2)
```

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-01-09 14:47:35+00:00_

Thanks!

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 14:59:34+00:00_

Closes #1070: Avoid un-pickling issues under Python 3 when accessing doctrees saved under Python 2.

→ <<cset e26b02e23e29e033a6ab4747464502d087235792>>
